sympy/matrices/random.py

```diff
@@ -0,0 +1,2248 @@
+    [             0,             1, -1, 0]])
+
+    Parameters
+    ==========
+    mat : Matrix
+        a complex matrix
+
+    """
+    dim = _shape(mat)[0]
+    mat = mat or identity_matrix(dim)
+    obj = super_elementary_matrix(2 * dim)
+    for i in range(dim):
+        for j in range(dim):
+            z_value = _get_value(mat, i, j)
+            a, b = _re(z_value), _im(z_value)
+            _set_value(obj, 2 * i, 2 * j, a)
+            _set_value(obj, 2 * i, 2 * j + 1, b)
+            _set_value(obj, 2 * i + 1, 2 * j, -b)
+            _set_value(obj, 2 * i + 1, 2 * j + 1, a)
+    return obj
+
+
+def regular_to_singular(mat, rank=None):
+    r""" build matrix with linear combination of matrix columns to meet rank
+
+    Explanation
+    ===========
+    Let $I$ be the $n \times n$ identity_matrix matrix
+    and $D$ an diagonal matrix with only zero and one entries
+    of rank $r$ (i.e. all entries but $n-r$ diagonal entries are zero).
+
+    If a full rank matrix $A$ is given, then
+
+    .. math::
+        B = D * (I + A * (I-D)) = D + D * A * (I-D)
+
+    is of rank $r$, too.  Some for $A*B$ which contains $r$ columns with $A$.
+
+    If $A$ is an upper triangular matrix $B$ as well $A*B$ is one, too.
+
+    Here, $A*B$ is returned.
+    Note, if $n=r$ the matrix $B$ is the identity_matrix.
+
+    Examples
+    ========
+    >>> from sympy import Matrix
+    >>> from sympy.matrices.random import regular_to_singular
+    >>> m = Matrix([[1, 2, 3],[4, 5, 6], [7, 8 ,0]])
+    >>> m.rank()
+    3
+    >>> n = regular_to_singular(m, 2)
+    >>> n
+    Matrix([
+    [1, 2, 15],
+    [4, 5, 42],
+    [7, 8, 69]])
+    >>> n.rank()
+    2
+
+    Parameters
+    ==========
+    mat : Matrix
+        a complex matrix
+    rank : integer
+        rank of matrix
+
+    """
+    if rank is None:
+        return mat
+    dim = _shape(mat)[0]
+    if rank == dim:
+        return mat
+    i = identity_matrix(dim)
+    p = permutation(dim)
+    d = projection(dim, (0, rank))
+    d = _multiply(p.inv(), d, p)
+    md = _multiply(mat, d)
+    return md + _multiply(md, mat, i - d)
+
+
+# === base matrices ===
+
+
+def identity_matrix(dim):
+    r"""identity_matrix matrix n x n
+
+    Explanation
+    ===========
+
+    Creates identity_matrix matrix with only ones on the diagonal
+    and zeros anywhere else.
+
+    Examples
+    ========
+
+    >>> from sympy.matrices.random import identity_matrix
+
+    >>> identity_matrix(3)
+    Matrix([
+    [1, 0, 0],
+    [0, 1, 0],
+    [0, 0, 1]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension n of matrix
+
+    See Also
+    ========
+    diagonal
+    sympy.matrices.dense.eye
+
+    """
+
+    return super_elementary_matrix(dim)
+
+
+def projection(dim,
+               index=None):
+    r"""a projection matrix n x n
+
+    Explanation
+    ===========
+    A projection is a identity_matrix like matrix
+    but with zero diagonal entiries off **index**.
```

Comment:
Single `*` like `*index*` for a parameter.

Should be a blank line under `=========`

Reply:
entries